src/VisualStudio/Core/Def/Implementation/ProjectSystem/CPS/IWorkspaceProjectContext.cs

```diff
@@ -58,5 +60,26 @@ internal interface IWorkspaceProjectContext : IDisposable
         void EndBatch();
 
         void ReorderSourceFiles(IEnumerable<string> filePaths);
+
+        /// <summary>
+        /// Applies changes to symbols for documents after their file path has changed. The same policy
+        /// is applied across all documents based on the <see cref="FilePathChangeSymbolChange"/>
+        /// </summary>
+        /// <param name="documentIdToOriginalPathMapping">Mapping of document ids to the original file path. Original path may be used to determine if symbol changes are applicable</param>
+        /// <remarks>
+        /// Note that calls to this may do expensive work to fix up symbol renaming. Symbol changes express a desire to change the symbol, but may not result in a symbol change if 
+        /// deemed unnecessary. For example, if no types matched the file name before then a named type change will not be performed. 
+        /// </remarks>
+        void ApplyDocumentSymbolChanges(ImmutableDictionary<DocumentId, string> documentIdToOriginalPathMapping, FilePathChangeSymbolChange filePathChangeSymbolChange, CancellationToken cancellationToken);
```

Comment:
What would this mean though for cross-project cases? I was imagining this would end up somewhere else...

Reply:
imo this still works for cross product, although maybe the naming suggests otherwise. Passing a value of `{ docId, null }` should be allowed, and we could handle that however we want to. That would cover remove/add across projects. Maybe I'm missing how hard that would be for the project system to coordinate?

Reply:
I guess I imagine the IWorkspaceProjectContext as really being scoped to a specific project for the most part. The component that actually holds onto this in CPS may not be aware of what's going on.

Reply:
Are you thinking something like VSWorkspace or even Workspace? That's a set of projects and some other information, so may be the right place. If so, what's the contract that we expose to the project system?

Reply:
Moved to VSWorkspace for now